pytorch_lightning/metrics/retrieval/retrieval_metric.py

```diff
@@ -0,0 +1,117 @@
+import torch
+from typing import List, Optional, Callable, Any
+
+from pytorch_lightning.metrics import Metric
+from pytorch_lightning.metrics.utils import get_mini_groups
+
+
+IGNORE_IDX = -100
+
+
+class RetrievalMetric(Metric):
+    r"""
+    Works with binary data. Accepts integer or float predictions from a model output.
+
+    Forward accepts
+    - ``indexes`` (long tensor): ``(N, ...)``
+    - ``preds`` (float tensor): ``(N, ...)``
+    - ``target`` (long or bool tensor): ``(N, ...)``
+
+    `indexes`, `preds` and `target` must have the same dimension.
+    `indexes` indicate to which query a prediction belongs.
+    Predictions will be first grouped by indexes and then the
+    actual metric will be computed as the mean of the scores over each query.
+
+    Args:
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects
+            the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When `None`, DDP
+            will be used to perform the allgather. default: None
+        query_without_relevant_docs:
+            Specify what to do with queries that do not have at least a positive target. Choose from:
+
+            - ``'skip'``: skip those queries (default)
+            - ``'error'``: raise a ``ValueError``
+            - ``'pos'``: score on those queries is counted as ``1.0``
+            - ``'neg'``: score on those queries is counted as ``0.0``
+
+        exclude:
+            Do not take into account predictions where the target is equal to this value. default: -100
+    """
+
+    query_without_relevant_docs_options = ['error', 'skip', 'pos', 'neg']
+
+    def __init__(
+        self,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        query_without_relevant_docs: str = 'skip',
```

Comment:
Just to be consistent with the remaining code base: move all parameters specific to the metric to be first, and then the four base metric arguments.